In [2]:
import pandas as pd
import numpy as np

# --- 1. Load the Dataset ---
# We're loading the dataset into a pandas DataFrame. Since it's an Excel file,
# we use pd.read_excel() and need the 'openpyxl' library installed.
# Using a try-except block is good practice to handle potential errors.

try:
    # The file path assumes the notebook is in a 'notebooks' folder
    # and the data is in a 'data' folder, both inside the project root.
    # Make sure your file has a .xlsx or .xls extension.
    df = pd.read_excel(r'C:\Users\jaiku\PycharmProjects\Ola_Ride_Analytics\data\OLA_DataSet.xlsx')
    print("✅ Excel file loaded successfully!")
except FileNotFoundError:
    print("❌ Error: 'ola_data.xlsx' not found. Make sure it's in the 'data' directory.")
    # Exit gracefully if the file isn't found
    exit()
except Exception as e:
    print(f"An error occurred: {e}")
    exit()


# --- 2. Initial Reconnaissance ---
# This step is like a doctor's initial check-up on a patient. We are getting
# a high-level overview of the dataset's health and structure.

print("\n--- First 5 Rows of the Dataset (df.head()) ---")
# df.head() shows us the first few rows. It helps us visually inspect the
# column names, the type of data in each column (text, numbers, dates?),
# and check for any obvious issues like formatting problems.
print(display(df.head()))


print("\n--- Technical Summary of the Dataset (df.info()) ---")
# df.info() gives a technical summary. We pay close attention to:
# - Non-Null Count: This immediately tells us which columns have missing data.
# - Dtype: This is the data type of each column. A column like 'Booking_Value'
#   being an 'object' instead of a number (int64/float64) is a red flag
#   that it contains non-numeric characters and needs cleaning.
print(df.info())


print("\n--- Statistical Summary of the Dataset (df.describe()) ---")
# df.describe(include='all') provides statistical summaries.
# For NUMERIC columns: look at 'min' and 'max' for impossible values
# (e.g., negative Ride_Distance, a rating above 5). Check the 'mean' vs.
# '50%' (median) to get a sense of skewness.
# For CATEGORICAL (object) columns: 'unique' tells us how many distinct
# categories exist. 'top' shows the most frequent category, and 'freq'
# shows its count. This is great for understanding the distribution.
print(display(df.describe(include='all')))

# --- 3. Initial Observations & Next Steps ---
# Based on the output, we can start forming a plan.
# For example, if 'Date' and 'Time' are 'object' types, we know our first
# cleaning task will be to combine and convert them to a proper datetime format.
# If there are many missing values in 'Customer_Rating', we need to investigate why.
print("\n--- Initial Analysis Complete ---")




✅ Excel file loaded successfully!

--- First 5 Rows of the Dataset (df.head()) ---


Date      Time     Booking_ID        Booking_Status  \
0 2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1 2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2 2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3 2024-07-22 03:15:00  03:15:00  CNR2395710036  Canceled by Customer   
4 2024-07-02 09:02:00  09:02:00  CNR1797421769               Success   

  Customer_ID Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  \
0   CID713523  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN   
1   CID225428         Bike     Magadi Road       Varthur  203.0   30.0   
2   CID270156    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0   
3   CID581320        eBike      HSR Layout   Vijayanagar    NaN    NaN   
4   CID939555         Mini     Rajajinagar   Chamarajpet  252.0   80.0   

                     Canceled_Rides_by_Customer      Canceled_Rides_by_Driver  \
0                                           NaN  Personal & Car related issue   
1                                           NaN                           NaN   
2                                           NaN                           NaN   
3  Driver is not moving towards pickup location                           NaN   
4                                           NaN                           NaN   

  Incomplete_Rides Incomplete_Rides_Reason  Booking_Value Payment_Method  \
0              NaN                     NaN            444            NaN   
1               No                     NaN            158           Cash   
2               No                     NaN            386            UPI   
3              NaN                     NaN            384            NaN   
4               No                     NaN            822    Credit Card   

   Ride_Distance  Driver_Ratings  Customer_Rating  \
0              0             NaN              NaN   
1             13             4.1              4.0   
2             40             4.2              4.8   
3              0             NaN              NaN   
4             45             4.0              3.0   

                                      Vehicle Images  
0  https://cdn-icons-png.flaticon.com/128/14183/1...  
1  https://cdn-icons-png.flaticon.com/128/9983/99...  
2  https://cdn-icons-png.flaticon.com/128/9983/99...  
3  https://cdn-icons-png.flaticon.com/128/6839/68...  
4  https://cdn-icons-png.flaticon.com/128/3202/32...

None

--- Technical Summary of the Dataset (df.info()) ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 20 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Date                        103024 non-null  datetime64[ns]
 1   Time                        103024 non-null  object        
 2   Booking_ID                  103024 non-null  object        
 3   Booking_Status              103024 non-null  object        
 4   Customer_ID                 103024 non-null  object        
 5   Vehicle_Type                103024 non-null  object        
 6   Pickup_Location             103024 non-null  object        
 7   Drop_Location               103024 non-null  object        
 8   V_TAT                       63967 non-null   float64       
 9   C_TAT                       63967 non-null   float64       
 10  Canceled_Rides_by_Customer  10499 non-null   

Date      Time     Booking_ID Booking_Status  \
count                          103024    103024         103024         103024   
unique                            NaN      1440         103024              4   
top                               NaN  00:53:00  CNR7153255142        Success   
freq                              NaN       101              1          63967   
mean    2024-07-16 11:31:38.879678720       NaN            NaN            NaN   
min               2024-07-01 00:00:00       NaN            NaN            NaN   
25%               2024-07-08 18:41:00       NaN            NaN            NaN   
50%               2024-07-16 11:23:00       NaN            NaN            NaN   
75%               2024-07-24 05:18:00       NaN            NaN            NaN   
max               2024-07-31 23:58:00       NaN            NaN            NaN   
std                               NaN       NaN            NaN            NaN   

       Customer_ID Vehicle_Type Pickup_Location Drop_Location         V_TAT  \
count       103024       103024          103024        103024  63967.000000   
unique       94544            7              50            50           NaN   
top      CID954071  Prime Sedan    Banashankari        Peenya           NaN   
freq             5        14877            2201          2159           NaN   
mean           NaN          NaN             NaN           NaN    170.876952   
min            NaN          NaN             NaN           NaN     35.000000   
25%            NaN          NaN             NaN           NaN     98.000000   
50%            NaN          NaN             NaN           NaN    168.000000   
75%            NaN          NaN             NaN           NaN    238.000000   
max            NaN          NaN             NaN           NaN    308.000000   
std            NaN          NaN             NaN           NaN     80.803640   

               C_TAT                    Canceled_Rides_by_Customer  \
count   63967.000000                                         10499   
unique           NaN                                             5   
top              NaN  Driver is not moving towards pickup location   
freq             NaN                                          3175   
mean       84.873372                                           NaN   
min        25.000000                                           NaN   
25%        55.000000                                           NaN   
50%        85.000000                                           NaN   
75%       115.000000                                           NaN   
max       145.000000                                           NaN   
std        36.005100                                           NaN   

            Canceled_Rides_by_Driver Incomplete_Rides Incomplete_Rides_Reason  \
count                          18434            63967                    3926   
unique                             4                2                       3   
top     Personal & Car related issue               No         Customer Demand   
freq                            6542            60041                    1601   
mean                             NaN              NaN                     NaN   
min                              NaN              NaN                     NaN   
25%                              NaN              NaN                     NaN   
50%                              NaN              NaN                     NaN   
75%                              NaN              NaN                     NaN   
max                              NaN              NaN                     NaN   
std                              NaN              NaN                     NaN   

        Booking_Value Payment_Method  Ride_Distance  Driver_Ratings  \
count   103024.000000          63967  103024.000000    63967.000000   
unique            NaN              4            NaN             NaN   
top               NaN           Cash            NaN             NaN   
freq         

None

--- Initial Analysis Complete ---


## Analysis of the Initial Output

You've successfully loaded the data, and the initial outputs from `df.info()` and `df.describe()` are revealing several important characteristics of your dataset.

### Key Insights from df.info()

- Missing Data is Structural, Not Random: Notice that several columns (`V_TAT`, `C_TAT`, `Payment_Method`, `Driver_Ratings`, `Customer_Rating`) have the exact same number of non-null entries: 63,967. The `df.describe()` output shows that there are also exactly 63,967 `Success` bookings. This is a crucial insight! It tells us that this data is only recorded for successful rides. This isn't messy data; it's a business rule. We should not try to "fill in" these missing values for cancelled rides.
- Cancellation Reasons are Separate: The columns `Canceled_Rides_by_Customer` and `Canceled_Rides_by_Driver` have many nulls because a ride can only be cancelled by one party. They are mutually exclusive.
- Incorrect Data Types: The `Time` column is an object (text), and more importantly, it's redundant since the `Date` column already includes the full timestamp. `Incomplete_Rides` is also an object when it should logically be a Yes/No or True/False value.
- Useless Column: The `Vehicle Images` column contains image URLs. While interesting, they are not useful for our analytical goals and can be removed.

### Key Insights from df.describe()

- Ride Distance Anomaly: The min value for `Ride_Distance` is 0. A ride with zero distance is physically impossible and needs to be investigated. These are likely records of bookings that were cancelled immediately after being accepted.
- Booking Value Skewness: The mean booking value (548) is significantly higher than the median (50% percentile) value (386). This indicates that the data is right-skewed. In simple terms, there are a few very expensive rides (likely long distances or high surge pricing) that are pulling the average up.
- Clean Ratings Data: The `Driver_Ratings` and `Customer_Rating` columns have a min of 3.0 and max of 5.0. This looks like very clean and well-structured data with no obvious errors.
- Categorical Dominance: For `Booking_Status`, `Success` is the most frequent category. For `Payment_Method`, `Cash` is the most popular. For cancellation reasons, "Driver is not moving" (customer) and "Personal & Car related issue" (driver) are the most common. These are valuable initial business insights.




### Key Columns Summary

| Column | Data Type | Issue/Insight | Action Needed |
|--------|-----------|---------------|---------------|
| `V_TAT`, `C_TAT`, `Payment_Method` | Mixed | Only populated for successful rides (63,967 entries) | Keep as-is, structural missing data |
| `Driver_Ratings`, `Customer_Rating` | Float | Clean data, range 3.0-5.0 | No action needed |
| `Canceled_Rides_by_Customer/Driver` | Object | Mutually exclusive, many nulls | Keep as-is, business logic |
| `Time` | Object | Redundant with `Date` column | Consider removing |
| `Incomplete_Rides` | Object | Should be boolean | Convert to True/False |
| `Vehicle Images` | Object | URLs, not analytical | Remove column |
| `Ride_Distance` | Float | Contains impossible 0 values | Investigate and clean |
| `Booking_Value` | Float | Right-skewed distribution | Consider log transformation |
| `Booking_Status` | Object | Dominated by "Success" | No action needed |

This analysis provides a solid foundation for your data cleaning and preprocessing steps.